tests/test-cluster.py

```diff
@@ -463,3 +465,121 @@ def test_no_cert_reissue_in_nodes(self):
+
+            if minor == "0" and major >= "1":
+                major = str(int(major) - 1)
+                minor = "9"
+            else:
+                if "-" in track:
+                    minor = str(int(minor.split("-")[0]) - 1)
+                else:
+                    minor = str(int(minor) - 1)
+
+            branch = "-" + branch if branch != "" else ""
+            older_version = major + "." + minor + branch + "/" + "stable"
+            print("Old version is {}".format(older_version))
+
+            type(self).VM = []
+            if not reuse_vms:
+                print("Creating machine")
+                vm = VM(backend)
+                vm.setup(older_version)
+                print("Waiting for machine")
+                vm.run("/snap/bin/microk8s.status --wait-ready --timeout 120")
+                self.VM.append(vm)
+            else:
+                vm = VM(backend, reuse_vms[0])
+                vm.setup(older_version)
+                self.VM.append(vm)
+
+            vm_older_version = self.VM[0]
+
+            # Wait for CNI pods
+            print("Waiting for cni")
+            while True:
+                ready_pods = 0
+                pods = vm_older_version.run(
+                    "/snap/bin/microk8s.kubectl get po -n kube-system -o wide"
+                )
+                for line in pods.decode().splitlines():
+                    if "calico" in line and "Running" in line:
+                        ready_pods += 1
+                if ready_pods == (len(self.VM) + 1):
+                    print(pods.decode())
+                    break
+                time.sleep(5)
+            yield
+
+        finally:
+            print("Cleanup up cluster")
+            if not reuse_vms:
+                for vm in self.VM:
+                    print("Releasing machine {} in {}".format(vm.vm_name, vm.backend))
+                    vm.release()
+
+    def test_mixed_version_join(self):
+        """
+        Test n versioned node joining a n-1 versioned cluster.
+        """
+        print("Setting up an newer versioned node")
+        vm = VM()
+        vm.setup(channel_to_test)
+        self.VM.append(vm)
+
+        # Form cluster
+        vm_older_version = self.VM[0]
+        print("Adding newer versioned machine {} to cluster".format(vm.vm_name))
+        add_node = vm_older_version.run("/snap/bin/microk8s.add-node")
+        endpoint = [ep for ep in add_node.decode().split() if ":25000/" in ep]
+        vm.run("/snap/bin/microk8s.join {}".format(endpoint[0]))
+
+        # Wait for nodes to be ready
+        print("Waiting for two node to be Ready")
+        attempt = 0
+        while attempt < 10:
+            try:
+                connected_nodes = vm_older_version.run("/snap/bin/microk8s.kubectl get no")
+                num_nodes = connected_nodes.count(b" Ready")
+                if num_nodes != 2:
+                    time.sleep(5)
+                    continue
```

Comment:
Shouldn't we increase the `attempts` counter here?

Reply:
The attempt counter is for how many times there was an error while listing the nodes rather than how many times there was an attempt to list the nodes.
We do a similar thing for other tests as well.

Reply:
Sure but if for some reason we break the test and the node does not get to Ready the test will just hung instead of timeout with an error. Lets fail after 5 mins or something like that.

Reply:
Fixed.